src/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,371 @@
+use float_cmp::{ApproxEq, F64Margin, approx_eq};
+
+use super::vector3::Vector3;
+
+
+/// A Rust implementation of Quaternion, we use this instead of existing packages
+/// because we want a C-safe representational structure. These quaternions use
+/// the Real-I-J-K standard, so those using JPL or other standards should take
+/// care to convert before initializing a Quaternion struct from the library
+#[repr(C)]
+#[derive(Copy, Clone, Debug)]
+pub struct Quaternion {
+    pub real: f64,
+    pub i: f64,
+    pub j: f64,
+    pub k: f64,
+}
+
+impl Quaternion {
+    /// Initializes a quaternion from a real component and an
+    /// imaginary 3-vector
+    pub fn new_from_vector(real: f64, imag: Vector3) -> Self {
+        Self{real: real, i: imag.x, j: imag.y, k: imag.z}
+    }
+
+    /// Initializes a Quaternion from its raw components
+    pub fn new(real: f64, i: f64, j: f64, k: f64) -> Self {
+        Self{real, i, j, k}
+    }
+
+    /// Converts from euler angles to a quaternion. The euler angles are expected to
+    /// be represented according to the Tait-Bryan formalism and applied in the Z-Y'-X"
+    /// order (where Z -> yaw, Y -> pitch, X -> roll)
+    pub fn from_euler_angles(roll: f64, pitch: f64, yaw: f64) -> Self {
+        let roll_cos = (roll * 0.5).cos();
+        let roll_sin = (roll * 0.5).sin();
+
+        let pitch_cos = (pitch * 0.5).cos();
+        let pitch_sin = (pitch * 0.5).sin();
+
+        let yaw_cos = (yaw * 0.5).cos();
+        let yaw_sin = (yaw * 0.5).sin();
+
+        let real = (roll_cos * pitch_cos * yaw_cos) + (roll_sin * pitch_sin * yaw_sin);
+        let i = (roll_sin * pitch_cos * yaw_cos) - (roll_cos * pitch_sin * yaw_sin);
+        let j = (roll_cos * pitch_sin * yaw_cos) + (roll_sin * pitch_cos * yaw_sin);
+        let k = (roll_cos * pitch_cos * yaw_sin) - (roll_sin * pitch_sin * yaw_cos);
+
+        Quaternion::new(real, i, j, k)
+    }
+
+    /// Converts a quaternion into euler angles. The euler angles are 
+    /// represented according to the Tait-Bryan formalism and applied 
+    /// in the Z-Y'-X" order (where Z -> yaw, Y -> pitch, X -> roll). 
+    /// The return value is a list of [roll, pitch, yaw]
+    pub fn to_euler_angles(&self) -> [f64;3] {
```

Comment:
It may be worthwhile to allow the user to specify the rotation order for to/from euler angles.

https://github.com/go-gl/mathgl/blob/v1.0.0/mgl64/quat.go#L265

Reply:
Specifying the rotation order is just a permutation at the very end (or very beginning), yeah? That sounds like a nice-to-have, but not something for the MVP. I wouldn't write anything more than a TODO for it yet (but am open to being wrong!).

Reply:
we [don't currently though](https://github.com/viamrobotics/rdk/blob/main/spatialmath/quaternion.go#L49), is there a reason for that?

Reply:
any chance we can leave this as a follow-up? I'm pretty sure the most popular representation is ZYX

Reply:
The reason we don't currently is because the library I linked exists and does it :)
The other two common (at least, I've seen them be in use) representations are XYZ and ZYZ.
But I don't think it needs to gate this PR, as long as it goes on a to-do list (or a library that already implements this is found)